keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+     // is delayed until a Pod gets scheduled that needs the
+     // ResourceClaim. The scheduler will consider all
+     // resource requirements of that Pod and trigger
+     // allocation for a node that fits the Pod.
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the PotentialNodes field triggers a check
+   // in the driver on which of those nodes the resource
+   // might be made available. It then excludes nodes
+   // by listing those where that is not the case in
+   // UnsuitableNodes.
+   //
+   // Unsuitable nodes will be ignored by the scheduler
+   // when selecting a node for a Pod. All other nodes are
+   // potential candidates, either because no information
+   // is available yet or because allocation might
+   // succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically and/or after
+   // changing resource allocation for some other
+   // ResourceClaim until a node gets selected by
+   // the scheduler.
+   UnsuitableNodes []string
+
+   // This field will get set by the resource driver to
+   // inform the scheduler where it can schedule Pods
+   // using the ResourceClaim.
+   //
+   // A resource driver may already set this before
+   // the resource is allocated. The scheduler will
+   // then check this field in addition to UnsuitableNodes
+   // to filter out nodes where the resource cannot
+   // be allocated.
+   //
+   // For an allocated resource this field defines
+   // where the resource can be used by Pods.
+   //
+   // Setting this field is optional. If nil, the
+   // resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
```

Comment:
I assume this is for network/fabric use cases, right?  I am struggling with this a little bit because the node labels may change, and so I wonder how restrictive this selector can become without being a security risk.

Reply:
Volumes use the same approach. I can see how a user can shoot themselves into the foot by editing labels. Resource drivers need to handle cases where they are asked to prepare a resource on a node that wasn't meant to have access to it.